### PR TITLE
CORDA-904 - Fix evolver to work with setter instantiated classses

### PR DESCRIPTION
**Pull request by fenryka**

### CODE REVIEW

File: node-api/src/main/kotlin/net/corda/nodeapi/internal/serialization/amqp/EvolutionSerializer.kt [hunk 1]
```diff
@@ -96,45 +125,44 @@ class EvolutionSerializer(
         fun make(old: CompositeType, new: ObjectSerializer,
                  factory: SerializerFactory): AMQPSerializer<Any> {
 
-            val readersAsSerialized = linkedMapOf<String, OldParam>(
-                    *(old.fields.map {
-                        val returnType =  try {
+            // The order in which the properties where serialised is important and must be preserved
```
Comment (rick-r3) [inline node-api/src/main/kotlin/net/corda/nodeapi/internal/serialization/amqp/EvolutionSerializer.kt]: where --> were

File: node-api/src/main/kotlin/net/corda/nodeapi/internal/serialization/amqp/EvolutionSerializer.kt [hunk 2]
```diff
@@ -177,11 +235,19 @@ class EvolutionSerializerGetter : EvolutionSerializerGetterBase() {
...
-            factory.getSerializersByDescriptor().computeIfAbsent(typeNotation.descriptor.name!!) {
-                when (typeNotation) {
-                    is CompositeType -> EvolutionSerializer.make(typeNotation, newSerializer as ObjectSerializer, factory)
-                    is RestrictedType -> EnumEvolutionSerializer.make(typeNotation, newSerializer, factory, schemas)
+                                        schemas: SerializationSchemas): AMQPSerializer<Any> {
+        return factory.getSerializersByDescriptor().computeIfAbsent(typeNotation.descriptor.name!!) {
+            when (typeNotation) {
+                is CompositeType -> EvolutionSerializer.make(typeNotation, newSerializer as ObjectSerializer, factory)
+                is RestrictedType -> {
+                    // Due to fixes (historic but also potentially future) in out fingerprinter if we can alter
```
Comment (rick-r3) [inline node-api/src/main/kotlin/net/corda/nodeapi/internal/serialization/amqp/EvolutionSerializer.kt]: I think I was just reviewing a PR with better wording here.  Make sure this doesn't overwrite it.  Probably needs rebase

Comment (fenryka) [reply node-api/src/main/kotlin/net/corda/nodeapi/internal/serialization/amqp/EvolutionSerializer.kt]: done

File: node-api/src/main/kotlin/net/corda/nodeapi/internal/serialization/amqp/EvolutionSerializer.kt (hunk 1 above)
Comment (fenryka) [reply node-api/src/main/kotlin/net/corda/nodeapi/internal/serialization/amqp/EvolutionSerializer.kt]: done

**Comment by fenryka** [comment]
@rick-r3 changes made and rebased

File: node-api/src/main/kotlin/net/corda/nodeapi/internal/serialization/amqp/EvolutionSerializer.kt [hunk 3]
```diff
@@ -82,6 +80,37 @@ class EvolutionSerializer(
...
+
+            constructor.parameters.withIndex().forEach {
+                readersAsSerialized.get(it.value.name!!)?.apply {
+                    this.resultsIndex = it.index
+                } ?: if (!it.value.type.isMarkedNullable) {
+                    throw NotSerializableException(
+                            "New parameter ${it.value.name} is mandatory, should be nullable for evolution to worK")
+                }
+            }
+
```
Comment (rick-r3) [inline node-api/src/main/kotlin/net/corda/nodeapi/internal/serialization/amqp/EvolutionSerializer.kt]: Remove this blank line

File: node-api/src/main/kotlin/net/corda/nodeapi/internal/serialization/amqp/EvolutionSerializer.kt [hunk 4]
```diff
@@ -96,45 +125,44 @@ class EvolutionSerializer(
...
 
-            val readersAsSerialized = linkedMapOf<String, OldParam>(
-                    *(old.fields.map {
-                        val returnType =  try {
+            // The order in which the properties were serialised is important and must be preserved
+            val readersAsSerialized = LinkedHashMap<String, OldParam>()
+            old.fields.forEach {
+                readersAsSerialized[it.name] = try {
+                    OldParam(-1, PropertySerializer.make(it.name, EvolutionPropertyReader(),
                             it.getTypeAsClass(factory.classloader)
```
Comment (rick-r3) [inline node-api/src/main/kotlin/net/corda/nodeapi/internal/serialization/amqp/EvolutionSerializer.kt]: The formatting of these lines could be cleaned up.  The line above ends in `,` and then the line below begins with `,` and this one has no commas.  Let's put them on the ends of line.

File: node-api/src/main/kotlin/net/corda/nodeapi/internal/serialization/amqp/EvolutionSerializer.kt [hunk 5]
```diff
@@ -151,7 +179,37 @@ class EvolutionSerializer(
...
+                "Failed to instantiate instance of object $clazz")
+
+        // *must* read all the parameters in the order they were serialized
+        oldReaders.values.zip(obj).forEach {
+            // if that property still exists on the new object then set it
+            it.first.property.readProperty(it.second, schemas, input).apply {
+                setters[it.first.property.name]?.set(instance, this)
+            }
+        }
+
```
Comment (rick-r3) [inline node-api/src/main/kotlin/net/corda/nodeapi/internal/serialization/amqp/EvolutionSerializer.kt]: We can get rid of this blank line

File: node-api/src/main/kotlin/net/corda/nodeapi/internal/serialization/amqp/PropertySerializers.kt [hunk 6]
```diff
@@ -68,6 +75,22 @@ class PrivatePropertyReader(val field: Field, parentType: Type) : PropertyReader
...
+ * it explicit that no properties are ever actually read from an object as the evolution
+ * serializer should only be accessing the already serialized form.
+ */
+class EvolutionPropertyReader : PropertyReader() {
+    override fun read(obj: Any?): Any? {
+        throw UnsupportedOperationException("It should be impossible for an evolution serializer to "
+                + "be reading from an object")
+    }
+
+    override fun isNullable(): Boolean {
```
Comment (rick-r3) [inline node-api/src/main/kotlin/net/corda/nodeapi/internal/serialization/amqp/PropertySerializers.kt]: We can write this all on one line.

File: node-api/src/main/kotlin/net/corda/nodeapi/internal/serialization/amqp/EvolutionSerializer.kt (hunk 3 above)
Comment (fenryka) [reply node-api/src/main/kotlin/net/corda/nodeapi/internal/serialization/amqp/EvolutionSerializer.kt]: done

File: node-api/src/main/kotlin/net/corda/nodeapi/internal/serialization/amqp/EvolutionSerializer.kt (hunk 4 above)
Comment (fenryka) [reply node-api/src/main/kotlin/net/corda/nodeapi/internal/serialization/amqp/EvolutionSerializer.kt]: done

File: node-api/src/main/kotlin/net/corda/nodeapi/internal/serialization/amqp/EvolutionSerializer.kt (hunk 5 above)
Comment (fenryka) [reply node-api/src/main/kotlin/net/corda/nodeapi/internal/serialization/amqp/EvolutionSerializer.kt]: done

File: node-api/src/main/kotlin/net/corda/nodeapi/internal/serialization/amqp/PropertySerializers.kt (hunk 6 above)
Comment (fenryka) [reply node-api/src/main/kotlin/net/corda/nodeapi/internal/serialization/amqp/PropertySerializers.kt]: done

**Comment by fenryka** [comment]
@rick-r3 formatting cleanup up, all those terrible and evil blank lines removed ;)

**Comment by fenryka** [comment]
build fails will be fixed once this is rebased onto master with https://github.com/corda/corda/pull/2464 included

**Comment by fenryka** [comment]
@rick-r3 the final issue with this PR was the case where we seem to be writing the Comparable interface into the schema, which we then try to evolve and fail because no constructor (and we changed fingerprinting so it thinks it needs it)

This makes encountering that a no-op, but whether we should be writing that into the schema.... I don't think we should (and need to look into why we are doing so). However, if it's a thing that can be done by accident then we should cope with that from a usability POV.

Have a look (I've dismissed the "OK" so you get a second opinion)... I'm inclined to want to fix the fact we  are writing it into the scheme as a separate PR

**Pull request merged**